test/tasks/dump.test.ts

```diff
@@ -0,0 +1,627 @@
+          ? balance.sub(1)
+          : balance.add(index),
+        // note: adding the index guarantees that all amounts are different.
+        // This is important since we are checking if the output has the correct
+        // order by sorting by this value, and equality might cause the sorting
+        // order to change.
+        boughtAmount: constants.WeiPerEther.mul(1009 % (byte + 1)).add(index),
+        decimals: (byte + index) % 30,
+        symbol: `DUMPED${index}`,
+      };
+    });
+    expect(pseudorandomTokenConfigs).to.have.length(32);
+
+    api.mock.ignoreExpectationOrder();
+    const dumpedTokens: MockContract[] = [];
+    for (const {
+      balance,
+      fee,
+      allowance,
+      boughtAmount,
+      symbol,
+      decimals,
+      index,
+    } of pseudorandomTokenConfigs) {
+      const dumped = await mockErc20(deployer);
+      await dumped.mock.symbol.returns(symbol);
+      await dumped.mock.decimals.returns(decimals);
+      await dumped.mock.balanceOf.withArgs(user.address).returns(balance);
+      await dumped.mock.allowance
+        .withArgs(user.address, allowanceManager)
+        .returns(allowance);
+      api.mock.getFee
+        .withArgs({
+          sellToken: dumped.address,
+          buyToken: to.address,
+          kind: OrderKind.SELL,
+          amount: balance,
+        })
+        .returns(BigNumber.from(fee));
+      if (!isIndexWithTooLargeFee(index)) {
+        api.mock.estimateTradeAmount
+          .withArgs({
+            sellToken: dumped.address,
+            buyToken: to.address,
+            kind: OrderKind.SELL,
+            amount: BigNumber.from(balance).sub(fee),
+          })
+          .returns(BigNumber.from(boughtAmount));
+      }
+      dumpedTokens.push(dumped);
+    }
+
+    const {
+      toToken,
+      transferToReceiver,
+      instructions,
+    } = await getDumpInstructions({
+      ...defaultDumpInstructions,
+      hasCustomReceiver: true,
+      dumpedTokens: dumpedTokens.map((t) => t.address).concat(to.address),
+      toTokenAddress: to.address,
+    });
+
+    expect(toToken.symbol).to.deep.equal("TOTOKEN");
+    expect(toToken.decimals).to.deep.equal(101);
+    expect(isNativeToken(toToken)).to.be.false;
+    expect((toToken as Erc20Token).address).to.deep.equal(to.address);
+
+    expect(transferToReceiver).not.to.be.undefined;
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const { amount, token } = transferToReceiver!;
+    expect(amount).to.deep.equal(toTokenBalance);
+    expect(token).to.deep.equal(toToken);
+
+    const successfulConfigs = pseudorandomTokenConfigs.filter(
+      (_, i) => !isIndexWithTooLargeFee(i),
+    );
+    expect(instructions)
+      .to.have.length(successfulConfigs.length)
+      .and.to.be.greaterThan(10);
+    successfulConfigs.sort((lhs, rhs) =>
+      lhs.boughtAmount.eq(rhs.boughtAmount)
+        ? 0
+        : lhs.boughtAmount.lt(rhs.boughtAmount)
+        ? -1
+        : 1,
+    );
```

Comment:
I am still interested as to why we have to explicitly define comparators for this and went looking a bit. Stumbled upon [this](https://stackoverflow.com/questions/39850339/how-to-sort-object-array-based-on-key-in-typescript) which I have translated into this suggestion. Not sure if it works, but it looks nicer.


```suggestion
    successfulConfigs.sort((lhs, rhs) => lhs.boughtAmount.localeCompare(rhs.boughtAmount));
```

Reply:
To strictly answer the question, because I don't know any better way. Some stumbling blocks:
- our objects can only be compared with their own custom comparison method and don't play nicely with builtins. Comparing with `<` converts the `BigNumber` to a string and string comparison is very bad for numbers, leading to a lexicographic comparison that seemingly works and so most of the times stays unnoticed for long until the inevitable issue. This is also the case for `localeCompare`, for example this test passes:
```
it.only("compares weirdly", () => {
  const bn10 = BigNumber.from(10);
  const bn90 = BigNumber.from(90);
  const bn100 = BigNumber.from(100);
  expect(bn10.toString().localeCompare(bn100.toString())).to.equal(-1);
  expect(bn90.toString().localeCompare(bn100.toString())).to.equal(+1);
});
```
- the standard way to do it for integers would simply be `(a, b) => a - b`. So what would be great would be something like `(a, b) => a.sub(b).sign()`, but there is no `sign()` function for a `BigNumber` or equivalent.
- if only `.sort()` is used, then the order is lexicographic.

In any case, I'd be wary of using anything with the name `locale` in it if not to format the final output to the user. I can easily imagine that some locales use Latin letters (er even Arabic numerals) but sorted in a different order in their alphabet, which means that depending on the locale the logic of the script is different.

Reply:
What we could do is defining a `bigNumberCompare(a,b)` function somewhere and always use it. I also don't like that we need to rewrite this code each time, it's very easy to use an incorrect comparison order, a mistaken sign, a `lte` instead of a `lt`, ...